hw/Makefile

```diff
@@ -130,17 +130,15 @@ $(ips_reg_rust_install): $(tock_rust_destdir)/%: $(REG_OUTPUT_SW_DIR)/% | $(tock
 clean-regs-header:
 	rm -r -f ${REG_OUTPUT_SW_DIR}
 
+top-hjson = $(PRJ_DIR)/hw/$*/data/$*.hjson
+
 top: $(tops_gen) $(tops_reg)
-$(tops_gen):
-	$(eval $@_TOP := $(strip $(foreach top,$(TOPS),$(findstring $(top),$@))))
-	${PRJ_DIR}/util/topgen.py -t ${PRJ_DIR}/hw/$($@_TOP)/data/$($@_TOP).hjson \
-		-o ${PRJ_DIR}/hw/$($@_TOP)/ ${toolflags}
-
-$(tops_reg):
-	$(eval $@_TOP := $(strip $(foreach top,$(TOPS),$(findstring $(top),$@))))
-	mkdir -p ${REG_OUTPUT_DV_DIR}/$($@_TOP)
-	${PRJ_DIR}/util/topgen.py -t ${PRJ_DIR}/hw/$($@_TOP)/data/$($@_TOP).hjson \
-		-r -o ${REG_OUTPUT_DV_DIR}/$($@_TOP) ${toolflags}
+$(tops_gen): %_gen:
+	${PRJ_DIR}/util/topgen.py -t $(top-hjson) -o ${PRJ_DIR}/hw/$*/ ${toolflags}
+
+$(tops_reg): %_reg:
```

Comment:
i'm a bit confused about this usage, does the `%_reg` mean even the ip reg targets will fall into this rule?

Reply:
ah..is it [this](https://www.gnu.org/software/make/manual/make.html#Static-Pattern)? I think my makefile knowledge just sucks.

Reply:
Yep, it's a static pattern. They're really neat (and much less confusing than implicit rules for exactly that reason!)